src/main/java/org/openrewrite/java/spring/boot3/MigrateWebMvcTagsToObservationConvention.java

```diff
@@ -0,0 +1,292 @@
+    private static final MethodMatcher TAGS_OF_ANY = new MethodMatcher("io.micrometer.core.instrument.Tags of(..)");
+    private static final MethodMatcher TAG_OF = new MethodMatcher("io.micrometer.core.instrument.Tag of(java.lang.String, java.lang.String)");
+    private static final MethodMatcher GET_TAGS = new MethodMatcher("* getTags(jakarta.servlet.http.HttpServletRequest, jakarta.servlet.http.HttpServletResponse, java.lang.Object, java.lang.Throwable)");
+
+    @Override
+    public String getDisplayName() {
+        return "Migrate `WebMvcTagsProvider` to `DefaultServerRequestObservationConvention`";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Migrate `WebMvcTagsProvider` to `DefaultServerRequestObservationConvention` as part of Spring Boot 3.2 removals.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(new UsesType<>(WEBMVCTAGSPROVIDER_FQ, true), new JavaVisitor<ExecutionContext>() {
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext ctx) {
+                J.MethodDeclaration getTagsMethod = null;
+                List<Statement> getTagsBodyStatements = new ArrayList<>();
+                for (Statement stmt : classDecl.getBody().getStatements()) {
+                    if (stmt instanceof J.MethodDeclaration) {
+                        J.MethodDeclaration md = (J.MethodDeclaration) stmt;
+                        if (GET_TAGS.matches(md, classDecl)) {
+                            getTagsMethod = md;
+                            if (getTagsMethod.getBody() != null) {
+                                getTagsBodyStatements.addAll(getTagsMethod.getBody().getStatements().subList(0, getTagsMethod.getBody().getStatements().size() - 1));
+                                Statement ret = getTagsMethod.getBody().getStatements().get(getTagsMethod.getBody().getStatements().size() - 1);
+                                if (ret instanceof J.Return && ((J.Return) ret).getExpression() instanceof J.MethodInvocation) {
+                                    if (TAGS_OF_ANY.matches(((J.Return) ret).getExpression())) {
+                                        getTagsBodyStatements.add((Statement) ((J.Return) ret).getExpression());
+                                    }
+                                }
+                                break;
+                            }
+
+                        }
+                    }
+                }
+
+                String tmpl = "class " + classDecl.getSimpleName() + " extends DefaultServerRequestObservationConvention {\n" +
+                              "    @Override\n" +
+                              "    public KeyValues getLowCardinalityKeyValues(ServerRequestObservationContext context) {\n" +
+                              "        KeyValues values = super.getLowCardinalityKeyValues(context);\n" +
```

Comment:
Did you explore whether to use [low cardinality versus high cardinality](https://docs.micrometer.io/micrometer/reference/observation/introduction.html#:~:text=High%20cardinality%20means%20that%20a,bounded%20number%20of%20possible%20values.)? It seems more risky to assume low cardinality. Any thoughts on that subject?